File: server/app/defense/policy.py
from __future__ import annotations

from typing import Any

from app.mail.interceptor import is_recipient_allowed


def evaluate_tool_policy(tool: str, args: dict[str, Any]) -> dict[str, Any]:
    if tool == "approve_candidate":
        return {
            "allow": False,
            "reason": "approve_candidate is disabled on Blue path — human recruiter must decide",
        }
    if tool == "send_email":
        to = str(args.get("to") or "")
        if not to:
            return {"allow": False, "reason": "missing recipient"}
        if not is_recipient_allowed(to):
            return {
                "allow": False,
                "reason": f"recipient domain not in EMAIL_DOMAIN_WHITELIST: {to}",
            }
        return {"allow": True, "reason": "recipient on whitelist"}
    return {"allow": False, "reason": "unknown tool"}

File: server/src/defense/policy.ts
import { isRecipientAllowed } from "../mail/interceptor.js";

export type ToolName = "send_email" | "approve_candidate";

export interface PolicyDecision {
  allow: boolean;
  reason: string;
}

export function evaluateToolPolicy(tool: ToolName, args: Record<string, unknown>): PolicyDecision {
  if (tool === "approve_candidate") {
    return {
      allow: false,
      reason: "approve_candidate is disabled on Blue path — human recruiter must decide",
    };
  }
  if (tool === "send_email") {
    const to = String(args.to || "");
    if (!to) return { allow: false, reason: "missing recipient" };
    if (!isRecipientAllowed(to)) {
      return { allow: false, reason: `recipient domain not in EMAIL_DOMAIN_WHITELIST: ${to}` };
    }
    return { allow: true, reason: "recipient on whitelist" };
  }
  return { allow: false, reason: "unknown tool" };
}
